certbot-apache/certbot_apache/_internal/prefetch_ocsp.py

```diff
@@ -0,0 +1,387 @@
+class DBMHandler(object):
+    """Context manager to handle DBM file reads and writes"""
+
+    def __init__(self, filename, mode):
+        self.filename = filename
+        self.filemode = mode
+        self.bsddb = False
+        self.database = None
+
+    def __enter__(self):
+        """Open the DBM file and return the filehandle"""
+
+        try:
+            import bsddb
+            self.bsddb = True
+            try:
+                self.database = bsddb.hashopen(self.filename, self.filemode)
+            except Exception:
+                raise errors.PluginError("Unable to open dbm database file.")
+        except ImportError:
+            # Python3 doesn't have bsddb module, so we use dbm.ndbm instead
+            import dbm
+            if self.filename.endswith(".db"):
+                self.filename = self.filename[:-3]
+            try:
+                self.database = dbm.ndbm.open(self.filename, self.filemode)  # pylint: disable=no-member
+            except Exception:
+                # This is raised if a file cannot be found
+                raise errors.PluginError("Unable to open dbm database file.")
+        return self.database
+
+    def __exit__(self, *args):
+        """Close the DBM file"""
+        if self.bsddb:
+            self.database.sync()
+        self.database.close()
+
+
+class OCSPPrefetchMixin(object):
+    """OCSPPrefetchMixin implements OCSP response prefetching"""
+
+    def __init__(self, *args, **kwargs):
+        self._ocsp_prefetch = {}  # type: Dict[str, str]
+        # This is required because of python super() call chain.
+        # Additionally, mypy isn't able to figure the chain out and needs to be
+        # disabled for this line. See https://github.com/python/mypy/issues/5887
+        super(OCSPPrefetchMixin, self).__init__(*args, **kwargs)  # type: ignore
+
+    def _ensure_ocsp_dirs(self):
+        """Makes sure that the OCSP directory paths exist."""
+        ocsp_work = os.path.join(self.config.work_dir, "ocsp_work")
+        ocsp_save = os.path.join(self.config.work_dir, "ocsp")
+        for path in [ocsp_work, ocsp_save]:
+            if not os.path.isdir(path):
+                filesystem.makedirs(path, 0o755)
+
+    def _ensure_ocsp_prefetch_compatibility(self):
+        """Make sure that the operating system supports the required libraries
+        to manage Apache DBM files.
+
+        :raises: errors.NotSupportedError
+        """
+        try:
+            import bsddb  # pylint: disable=unused-variable
+        except ImportError:
+            import dbm
+            if not hasattr(dbm, 'ndbm') or dbm.ndbm.library != 'Berkeley DB':  # pylint: disable=no-member
+                msg = ("Unfortunately your operating system does not have a "
+                       "compatible database module available for managing "
+                       "Apache OCSP stapling cache database.")
+                raise errors.NotSupportedError(msg)
+
+    def _ocsp_refresh_needed(self, pf_obj):
```

Comment:
After talking to Max, we decided to land this in `master` despite this limitation, but in looking into my suggestion for how to deal with the inevitable API change, I think I found an easy way around this problem now.

The one type of "new style enhancement" hook that we don't currently define for the OCSP prefetch enhancement is a deploy hook and guess when that hook is called? [Right before we call `restart()` causing Apache to start using the new certificates.](https://github.com/certbot/certbot/blob/6df90d17aeb3865ceda8ae30372cb1416a1a6c3b/certbot/certbot/_internal/main.py#L1187) It seems like past you had thought about the potential need to do something like this and already set up hooks for it. Nice work :smile:

If I'm thinking about this properly, it seems like we could easily add a method like `deploy_ocsp_prefetch` whose implementation in the Apache plugin is essentially the same as `update_ocsp_prefetch` and it will solve this problem. If so, I think this should be done in this PR.

What do you think?